File: Leetcode/Subsets.java
// Leetcode 78. Subsets
// https://leetcode.com/problems/subsets/description/
// Runtime: 4ms

// 递归轮着往list里加数
class Solution {
    void subsetRecr(int[] nums, int idx, LinkedList<Integer> list, List<List<Integer>> res) {
        for (int j = idx + 1; j < nums.length; j++) {
            // 这部分应该可以再优化
            LinkedList ll = new LinkedList();
            ll.addAll(list);
            ll.add(nums[j]);
            res.add(ll);
            subsetRecr(nums, j, ll, res);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();

        // 先加一个empty set
        List<Integer> empty = new LinkedList<>();
        res.add(empty);

        if (nums.length == 0)
            return res;

        for (int i = 0; i < nums.length; i++) {
            LinkedList<Integer> ll = new LinkedList<>();
            ll.add(nums[i]);
            res.add(ll);
            subsetRecr(nums, i, ll, res);
        }
        return res;
    }
}

File: Leetcode/078. Subsets/subsets.cpp
// Leetcode 78. Subsets
// https://leetcode.com/problems/subsets/
// time complexity: O(n x 2 power of n)
// space complexity: O(n x 2 power of n)

class Solution 
{
public:
    vector<vector<int>> subsets(vector<int>& nums) 
    {
        vector<vector<int>> ret;
        ret.push_back(vector<int>());
        
        for (int n : nums)
        {
            int size = ret.size();
            for (int i = 0; i < size; i++)
            {
                vector<int> v = ret[i];
                v.push_back(n);
                ret.push_back(v);
            }
        }
        
        return ret;
    }
};
